packages/program/src/contract-call-script.ts

```diff
@@ -91,55 +120,112 @@ const getMainCallReceipt = (
       type === ReceiptType.Call && from === SCRIPT_WRAPPER_CONTRACT_ID && to === contractId
   );
 
-const scriptResultDecoder = (contractId: AbstractAddress) => (result: ScriptResult) => {
-  if (toNumber(result.code) !== 0) {
-    throw new Error(`Script returned non-zero result: ${result.code}`);
-  }
+const scriptResultDecoder =
+  (contractId: AbstractAddress, isOutputDataHeap: boolean) => (result: ScriptResult) => {
+    if (toNumber(result.code) !== 0) {
+      throw new Error(`Script returned non-zero result: ${result.code}`);
+    }
 
-  const mainCallResult = getMainCallReceipt(
-    result.receipts as TransactionResultCallReceipt[],
-    contractId.toB256()
-  );
-  const mainCallInstructionStart = bn(mainCallResult?.is);
-  const receipts = result.receipts as ReturnReceipt[];
-  return receipts
-    .filter(({ type, is }) => isReturnType(type) && mainCallInstructionStart.eq(bn(is)))
-    .map((receipt: ReturnReceipt) => {
-      if (receipt.type === ReceiptType.Return) {
-        return new U64Coder().encode((receipt as TransactionResultReturnReceipt).val);
-      }
-      if (receipt.type === ReceiptType.ReturnData) {
-        const encodedScriptReturn = arrayify(receipt.data);
-        return encodedScriptReturn;
-      }
+    const mainCallResult = getMainCallReceipt(
+      result.receipts as TransactionResultCallReceipt[],
+      contractId.toB256()
+    );
+    const mainCallInstructionStart = bn(mainCallResult?.is);
 
-      return new Uint8Array();
-    });
-};
+    const receipts = result.receipts as ReturnReceipt[];
+    return receipts
+      .filter(({ type }) => isReturnType(type))
+      .flatMap((receipt: ReturnReceipt, index, filtered) => {
+        if (!mainCallInstructionStart.eq(bn(receipt.is))) {
+          return [];
+        }
+        if (receipt.type === ReceiptType.Return) {
+          return [new U64Coder().encode((receipt as TransactionResultReturnReceipt).val)];
+        }
+        if (receipt.type === ReceiptType.ReturnData) {
+          const encodedScriptReturn = arrayify(receipt.data);
+          if (isOutputDataHeap && isReturnType(filtered[index + 1]?.type)) {
+            const nextReturnData: TransactionResultReturnDataReceipt = filtered[
+              index + 1
+            ] as TransactionResultReturnDataReceipt;
+            return concat([encodedScriptReturn, arrayify(nextReturnData.data)]);
+          }
+
+          return [encodedScriptReturn];
+        }
+
+        return [new Uint8Array()];
+      })
+      .filter((v) => v !== undefined);
+  };
 
 export const decodeContractCallScriptResult = (
   callResult: CallResult,
   contractId: AbstractAddress,
+  isOutputDataHeap = false,
```

Comment:
It feels weird to me to assign a default value to `isOutputDataHeap`. Why would `false` be the default and not `true`? If we take it an abstraction layer above and say that the signature of the function instead is:

```ts
export const decodeContractCallScriptResult = (
    callResult: CallResult,
    contractId: AbstractAddress,
    functionFragment: FunctionFragment,
): Uint8Array[] => {...}
```

we wouldn't be able to assign a default value to `FunctionFragment` the same way it was done to `isOutputDataHeap`... but the `isOutputDataHeap` value is actually received from `FunctionFragment` in the `InvocationResult.getDecodedValue` method.

It seems to me that the signature with `FunctionFragment` is truer to the use case, as `FunctionFragment` itself encapsulates the actual abi function based on which the decoding is happening.

Reply:
I meant more conceptually, IMO this information isn't something that `decodeContractCallScriptResult` should assert to have a default value, as it's working with `FunctionFragment`s (directly if my suggestion is taken or indirectly as it currently is in `InvocationResult.getDecodedValue`) who, in the end, are owners of that information. It should just say, "give me a `FunctionFragment` (or a boolean indicating that the output data type is a heap type), and I'll decode the result." If we left it as it is, one could provide a `callResult` where the data is actually a heap type and use the default `isOutputDataHeap = false` value, and the decoding would fail.  Default values are for situations where a function will work as expected both when the user provided their own value and when they used the default one (e.g. an `options` parameter).